qupath-extension-bioformats/src/main/java/qupath/lib/images/servers/bioformats/BioFormatsImageServer.java

```diff
@@ -1376,16 +1376,24 @@ BufferedImage openImage(TileRequest tileRequest, int series, int nChannels, bool
 				if (tileWidth <= 0 || tileHeight <= 0) {
 					throw new IOException("Unable to request pixels for region with downsampled size " + tileWidth + " x " + tileHeight);
 				}
-
-				// OME Zarr files provide z scaling (the number of z stacks decreases when the resolution becomes
-				// lower, like the width and height)
-				if (ipReader instanceof ZarrReader) {
-					z /= tileRequest.getDownsample();
-				}
 		
 				synchronized(ipReader) {
 					ipReader.setSeries(series);
 					ipReader.setResolution(level);
+
+					// Some files provide z scaling (the number of z stacks decreases when the resolution becomes
+					// lower, like the width and height), so z needs to be updated for levels > 0
+					if (level > 0) {
+						ipReader.setResolution(0);
+						int zStacksFullResolution = ipReader.getSizeZ();
+						ipReader.setResolution(level);
+						int zStacksCurrentResolution = ipReader.getSizeZ();
+
+						if (zStacksFullResolution != zStacksCurrentResolution) {
```

Comment:
Should this be `round` or `floor`? My intuition says `round` but my intuition isn't certain...

Also, it would be useful to look at the number of `z` slices for different levels in the sample images you have access to. It might be possible to compute the ratio between the `z` slices for a level and the zero level, to determine how much to rescale the `z` index (which might be more robust than using the `downsample` value, in case it's possible that the downsampling in `z` uses a different factor).

Reply:
I tried to use `round` but I got an error when `z / tileRequest.getDownsample()` was giving something like 133.5, which turned into 134 after applying `round`, when the max possible value was 133.